### PR TITLE
add Glm

### PR DESCRIPTION
**Pull request by Cyrilvallez**
GLM model!

### CODE REVIEW

File: src/transformers/models/glm/__init__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (ArthurZucker) [inline src/transformers/models/glm/__init__.py:1]:
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: src/transformers/models/glm/convert_glm_weights_to_hf.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    ".embedding.word_embeddings.": ".embed_tokens.",
+    ".encoder.final_layernorm.": ".norm.",
+    ".encoder.layers.": ".layers.",
+    "rotary_pos_embed.": "rotary_emb.",
+    "self_attention.": "self_attn.",
+    "query_key_value.": "qkv_proj.",
+    "dense.": "o_proj.",
+    "dense_h_to_4h.": "gate_up_proj.",
+    "dense_4h_to_h.": "down_proj.",
+}
```
Comment (ArthurZucker) [inline src/transformers/models/glm/convert_glm_weights_to_hf.py:24]: cool! Let's setup good standards however, see MLLAMA, full explicit regex are more informative IMO! 🤗

File: src/transformers/models/glm/convert_glm_weights_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        attention_dropout=original_config.pop("attention_dropout"),
+        max_position_embeddings=original_config.pop("seq_length"),
+        rms_norm_eps=original_config.pop("layernorm_epsilon"),
+        rope_theta=10000.0 * original_config.pop("rope_ratio", 1),
+        use_cache=original_config.pop("use_cache"),
+        head_dim=original_config.pop("kv_channels"),
+        attention_bias=original_config.pop("add_qkv_bias"),
+        eos_token_id=original_config.pop("eos_token_id"),
+        pad_token_id=original_config.pop("pad_token_id"),
+        tie_word_embeddings=original_config.pop("tie_word_embeddings"),
```
Comment (ArthurZucker) [inline src/transformers/models/glm/convert_glm_weights_to_hf.py:92]: Let's try to use ** here for attributes that have the same name

File: src/transformers/models/glm/modular_glm.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        super().__init__(config, layer_idx)
+        self.o_proj = nn.Linear(self.hidden_size, self.hidden_size, bias=False)
+        self.scaling = 1 / math.sqrt(self.head_dim)
+
+
+class GlmFlashAttention2(GlmAttention, GraniteFlashAttention2):
+    pass
+
+
+class GlmSdpaAttention(GlmAttention, GraniteSdpaAttention):
```
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:152]:
```suggestion
class GlmSdpaAttention(GraniteSdpaAttention):
```

Comment (ArthurZucker) [reply src/transformers/models/glm/modular_glm.py:152]: I think this should be enough

File: src/transformers/models/glm/modular_glm.py
Comment (ArthurZucker) [inline src/transformers/models/glm/modular_glm.py:1]: holy molly so nice!

File: tests/models/glm/test_modeling_glm.py [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+        output_fa_2 = model.generate(**inputs, max_new_tokens=15, do_sample=False)
+        output_fa_2 = tokenizer.batch_decode(output_fa_2)
+
+        self.assertListEqual(output_native, output_fa_2)
+
+    @parameterized.expand([("float16",), ("bfloat16",), ("float32",)])
+    @require_torch_sdpa
+    @slow
+    @is_flaky
+    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```
Comment (ArthurZucker) [inline tests/models/glm/test_modeling_glm.py:468]: why do we have to overwrite this one?

File: tests/models/glm/test_modeling_glm.py
Comment (ArthurZucker) [inline tests/models/glm/test_modeling_glm.py:1]: Cool! In general the least we have to overwrite the better!

Comment (ArthurZucker) [reply tests/models/glm/test_modeling_glm.py:1]: meaning are there ways to remove some of the tests you added?

**Review by ArthurZucker — COMMENTED** [review]
Very very nice!

File: src/transformers/models/glm/convert_glm_weights_to_hf.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/models/glm/convert_glm_weights_to_hf.py]: I didn't to avoid adding unused fields, but I refactored to make that block nicer to read.

File: tests/models/glm/test_modeling_glm.py (hunk 5 above)
Comment (Cyrilvallez) [reply tests/models/glm/test_modeling_glm.py:468]: Unfortunately, based on the random inputs there may be sometimes when one of the tests fail - this I overwrote it to add the `flaky` decorator which allows the test to consistently pass

File: tests/models/glm/test_modeling_glm.py
Comment (Cyrilvallez) [reply tests/models/glm/test_modeling_glm.py:1]: Unfortunately no -- based on the random seed, some are failing from time to time, and they need to be `flaky` to consistently pass

**Comment by Cyrilvallez** [comment]
Ready for last review @ArthurZucker, `setup_and_quality` fail because of the `__all__` issue, but will pass once https://github.com/huggingface/transformers/pull/33859 is merged.

**Comment by zRzRzRzRzRzRzR** [comment]
Thank you very much for your help. I also saw this [huggingface](https://huggingface.co/THUDM/glm-4-9b-chat/discussions/81) PR.
I have replied, and some of the code may need to be modified. Perhaps we can work together to improve it and merge this work.

Thank you again for your support!

**Comment by ArthurZucker** [comment]
Of course!

**Review by ArthurZucker — APPROVED** [review]
LGTM anything missing before we merge?

**Comment by Cyrilvallez** [comment]
> LGTM anything missing before we merge?

No, only issue are the docstrings in the configuration, but this will be solved with the auto-docstrings ; I will update it when it's ready

**Comment by liyucheng09** [comment]
@Cyrilvallez Hi Cyril, you PR for the 1M version of the model got an unexpected generation. Please refer to here for more information: [https://huggingface.co/THUDM/glm-4-9b-chat-1m/discussions/17](https://huggingface.co/THUDM/glm-4-9b-chat-1m/discussions/17).
